Implemented `build.build-dir` config option

ranger-ross opened the pull request:
## What does this PR try to resolve?

This PR adds a new `build.build-dir` configuration option that was proposed in https://github.com/rust-lang/cargo/issues/14125#issuecomment-2258708917

This new config option allows the user to specify a directory where intermediate build artifacts should be stored. 
I have shortened it to just `build-dir` from `target-build-dir`, although naming is still subject to change.

### What is a final artifact vs an intermediate build artifact

#### Final artifacts
These are the files that end users will typically want to access directly or indirectly via a third party tool.

* binary executable for bin crates
* binary executable for examples
* `.crate` files output from `cargo package`
* [depinfo files](https://doc.rust-lang.org/cargo/reference/build-cache.html#dep-info-files) (`.d` files) for third party build-system integrations (see https://github.com/rust-lang/cargo/blob/master/src/cargo/core/compiler/fingerprint/mod.rs#L194)
* `cargo doc` output (html/css/js/etc)
* Cargo's [`--timings`](https://doc.rust-lang.org/cargo/reference/timings.html) HTML report

#### Intermediate build artifact, caches, and state
These are files that are used internally by Cargo/Rustc during the build process

* other depinfo files (generated by rustc, fingerprint, etc. See https://github.com/rust-lang/cargo/blob/master/src/cargo/core/compiler/fingerprint/mod.rs#L164)
* rlibs and debug info from dependencies
* build script `OUT_DIR`
* output from proc macros (previously stored in `target/build`)
* [incremental build](https://doc.rust-lang.org/rustc/codegen-options/index.html#incremental) output from rustc
* fingerprint files used by Cargo for rebuild detection
* scratchpad used for `cargo package` verify step
* Cache of rustc invocations (`.rustc_info.json`)
* "pre and non uplifted" binary executables. (ie. bins for `examples` that contain the hash in the name, bins for `benches`, proc macros, build scripts)
* `CARGO_TARGET_TMPDIR` files (see rational for this [here](https://github.com/rust-lang/cargo/issues/14125#issuecomment-2258708917))
* [future-incompat-report](https://doc.rust-lang.org/cargo/reference/future-incompat-report.html)'s `.future-incompat-report.json` file

## Feature Gating Strategy

We are following the "Ignore the feature that is used without a gate" approach as described [here](https://doc.rust-lang.org/nightly/nightly-rustc/cargo/core/features/index.html).

The rational for this is: 
The `build.build-dir` is likely going to be set by by users "globally" (ie. `$CARGO_HOME/config.toml`) to set a shared build directory to reduce rebuilding dependencies. For users that multiple cargo versions having having an error would be disrupted.
The fallback behavior is to revert to the behavior of the current stable release (building in `$CARGO_TARGET_DIR`)

## Testing Strategy 

* We have the existing Cargo testsuite to be sure we do not introduce regressions.
  * I have also run the testsuite locally with the cli flag remove to verify all tests pass with the default build dir (which falls back to the target dir)
* For testing thus far, I have been using small hello world project with a few dependencies like `rand` to verify files are being output to the correct directory.
* When this PR is closer to merging, I plan to test with some larger projects with more dependencies, build scripts, ect.
* Other testing recommendations are welcome :bow: 

## How should we test and review this PR?

This is probably best reviewed commit by commit. I documented each commit. 
I tied to follow the atomic commits recommendation in the Cargo contributors guide, but I split out some commits for ease of review. (Otherwise I think this would have ended up being 1 or 2 large commits :sweat_smile:)


## Questions

- [x] What is the expected behavior of `cargo clean`? 
  * At the time of writing it only cleans `target` and does not impact the build-dir but this is easily changable.
  * Answer: See https://github.com/rust-lang/cargo/pull/15104#issuecomment-2616369862
- [x] When using `cargo package` are was expecting just the `.crate` file to be in `target` while all other output be stored in `build.build-dir`?  Not sure if we consider things like `Cargo.toml`, ` Cargo.toml.orig`, `.cargo_vcs_info.json` part of the user facing interface.
  * Current consensus is that only `.crate` is considered a final artifact 
- [x] Where should `cargo doc` output go? HTML/JS for many crates can be pretty large. Moving to the build-dir would help reduce duplication if we find the that acceptable. For `cargo doc --open` this is not a problem but may be problematic for other use cases?
  * Answer: https://github.com/rust-lang/cargo/pull/15104#issuecomment-2619760470
- [x] Are bins generated from `benches` considered final artifacts?
  * Since bins from `examples` are considered final artifacts, it seems natural that `benches` should also be considered final artifacts. However, unlike `examples` the `benches` bins are stored in `target/{profile}/deps` instead of a dedicated directory (like `target/{profile}/examples`). We could move them into a dedicated directory (`target/{profile}/benches`) but that mean would also be changing the structure of the `target` directory which feels out of scope for this change. If we decide that `benches` are final artifacts, it would probably be better to consider that changes as part of https://github.com/rust-lang/cargo/issues/6790
  * Answer: https://github.com/rust-lang/cargo/pull/15104#issuecomment-2636519011
- [ ] [Do we want to include a `CARGO_BUILD_DIR` shortcut env var?](https://github.com/rust-lang/cargo/pull/15104#discussion_r1943404348) 
  * The current commit (2af0c918d415b4de03e627459c3594826ae03cfb) has included the `CARGO_BUILD_DIR` shortcut. This can be removed before merging if there a good reason to.
 
## TODO

- Implementation 
  - [x] Add support in `cargo clean`
  - [ ] ~~Implement templating for `build.build-dir`~~
     * Will implement in a follow up PR [as suggested](https://github.com/rust-lang/cargo/pull/15104#issuecomment-2636514628)
  - [x] Fix issue with `target/examples` still containing "pre-uplifted" binaries
  - [x] Verify `build-dir` with non-bin crate types
- Prepare for review
  - [x] Clean up/improve docs
  - [x] Review tests and add more as needed
  - [x] Fix tests in CI (Windows is currently failing)
  - [x] Clean up commits
  - [ ] Resolve remaining questions
- [x] Request review

weihanglo commented on tests/testsuite/build_dir.rs:
We usually add some tests verifying a nightly flag is really gated.

weihanglo commented on src/cargo/core/compiler/build_runner/mod.rs:
```diff
@@ -290,7 +290,9 @@ impl<'a, 'gctx> BuildRunner<'a, 'gctx> {
 
-            super::output_depinfo(&mut self, unit)?;
+            if !self.bcx.gctx.cli_unstable().build_dir {
+                super::output_depinfo(&mut self, unit)?;
```
Why do we stop generating this when `build-dir` is enabled?

ranger-ross replied:
Sorry, actually I believe this is a mistake on my part :sweat_smile: 

My original assumption was that the `/target/debug/{project-name}.d` file was the same file as `/target/debug/deps/{project-name}-{hash}.d`, so I was trying to avoid copying it into `target`.
But now that I compared the files they appear to be different so it looks like that was a bad assumption.

I believe this should probably be considered an intermediate build artifact so it should probably be moved to the `build-dir` location. 

Modifying this file in particular was a bit tricky as it was not just replacing `.target_dir()` with `.build_dir()`.
I think I will probably need to go to where the [`OutputFile`](https://github.com/rust-lang/cargo/blob/e63457a4cecfc3562d87197c451a3a3146f5e1a6/src/cargo/core/compiler/build_runner/compilation_files.rs#L136) is created and modify the `hardlink` to be the new build-dir path.

ranger-ross replied:
Given [this](https://github.com/rust-lang/cargo/pull/15104#issuecomment-2619773382) comment I am going to revert this change

weihanglo reviewed (commented):
Just want to make sure I didn't miss something. From what I can tell these directories/files have been removed right?

* `target/<profile>/.metabuild`
* `target/<profile>/.fingerprint`
* `target/<profile>/deps`
* `target/<profile>/incremental`
* `target/<profile>/build`
* `target/.cargo-lock`
* `target/tmp`
* `target/.rustc_info.json`

ranger-ross commented:
> Just want to make sure I didn't miss something. From what I can tell these directories/files have been removed right?
> 
> * `target/<profile>/.metabuild`
> * `target/<profile>/.fingerprint`
> * `target/<profile>/deps`
> * `target/<profile>/incremental`
> * `target/<profile>/build`
> * `target/.cargo-lock`
> * `target/tmp`
> * `target/.rustc_info.json`

Yes, that with the exception of `target/.cargo-lock`. 
I think we will still want this cargo lock for backwards compatibility with previous versions of cargo.

Ideally in the longer term it can be removed in favor of fine grain locking like https://github.com/rust-lang/cargo/issues/4282

So a typical `target` directory will be something like 
```
target
├── CACHEDIR.TAG
└── debug
    ├── .cargo-lock
    ├── examples
    └── hello_world // (the binary)
```

epage commented on src/doc/src/reference/unstable.md:
````diff
@@ -237,6 +238,22 @@ This can also be specified in `.cargo/config.toml` files.
 artifact-dir = "out"
 ```
 
+## build-dir
+* Original Issue: [#14125](https://github.com/rust-lang/cargo/issues/14125)
+* Tracking Issue: [#14125](https://github.com/rust-lang/cargo/issues/14125)
+
+This feature allows you to specify the directory where intermediate build artifacts will be stored.
+The final artifacts will also be contained in this directory but will be hardlinked into the artifact directory (usually `target`)
+
+This can be specified in `.cargo/config.toml` files using the `build.build-dir`.
+
+Example:
+
+```toml
+[build]
+build-dir = "out"
+```
````
Could you include documentation in this that mirrors what would be in https://doc.rust-lang.org/cargo/reference/config.html

epage commented on src/cargo/core/compiler/layout.rs:
```diff
@@ -135,6 +135,12 @@ pub struct Layout {
     /// The lockfile for a build (`.cargo-lock`). Will be unlocked when this
     /// struct is `drop`ped.
     _lock: FileLock,
+    /// Same as `_lock` but for the build directory.
+    ///
+    /// NOTE: `_lock` and `_build_lock` can eventually be merged once #14125 is stablized.
+    /// Will be `None` when the build-directory and target-directory are the same path as we cannot
+    /// lock the same path twice.
```
How can we merge these locks?

In the short term, we'll have
- target-dir
- build-dir

which both need locking

In the long term, we'll have
- artifact-dir
- build-dir

which will both need locking

ranger-ross replied:
Sorry, you are right. Not quiet sure what I was thinking when I wrote this comment.

epage commented on tests/testsuite/build_dir.rs:
Also, we generally add (passing) tests before the change so the affect of the change is obvious.

This is more straightforward when fixing a bug.  In a case like this, a common option is to write the tests without the new feature or using something similar ( like `target-dir`) and then in the commit where the feature is added, you switch over the tests.

epage commented:
> What is the expected behavior of cargo clean? 

It should clean the build dir

epage commented:
> When using cargo package are was expecting just the .crate file to be in target while all other output be stored in build.build-dir? Not sure if we consider things like Cargo.toml,  Cargo.toml.orig, .cargo_vcs_info.json part of the user facing interface.

imo The artifact for `cargo package` is the `.crate`.  Everything else is part of the "build" process.

epage commented:
Can we call out explicitly what our testing strategy is?

We likely should also explicitly document in the PR what is considered an artifact and what is a build output and make sure we have tests for these.

ranger-ross commented:
One other question that came to my mind was the output of `cargo doc`. HTML/JS for many crates can be pretty large. Moving to the build-dir would help reduce duplication if we find the that acceptable. For `cargo doc --open` this is not a problem but may be problematic for other use cases?

Perhaps symlinking the `index.html` from the build dir into `target` could be an option if we care about keeping an entry point.

ranger-ross commented:
> Can we call out explicitly what our testing strategy is?
> 
> 
> 
> We likely should also explicitly document in the PR what is considered an artifact and what is a build output and make sure we have tests for these.


@epage sure, I updated the PR description but let me know if I missed anything.

epage commented:
> One other question that came to my mind was the output of cargo doc. HTML/JS for many crates can be pretty large. Moving to the build-dir would help reduce duplication if we find the that acceptable. For cargo doc --open this is not a problem but may be problematic for other use cases?

imo `cargo doc`s output is an artifact that people will want access to.  I suspect it'd be a breaking change to move it out of `target-dir`.

epage commented:
> depinfo files (.d files)

There are multiple types of depinfo files.  I suspect the ones next to final artifacts are also considered final artifacts, see https://doc.rust-lang.org/cargo/reference/build-cache.html#dep-info-files

epage commented:
FYI I added to the PR description a couple more intermediate artifacts

> * rlibs and debug info from dependencies
> * build script `OUT_DIR`

When are workspace member rlibs considered final artifacts?  We're putting them in `target/<profile>` at times, so I take it that has already been answered.

epage commented:
> > Can we call out explicitly what our testing strategy is?
> > We likely should also explicitly document in the PR what is considered an artifact and what is a build output and make sure we have tests for these.
> 
> @epage sure, I updated the PR description but let me know if I missed anything.

Sorry if I wasn't clear but my discussion of testing strategy was in the context of tracking the categorization of artifacts.  I'm not saying we have to exhaustively test it but we should think about it and document the choice made and why.

ranger-ross commented:
I spent some more time this weekend making some progress on this PR.

Here are my updates

* Fixed some issues to make `cargo clean` & `cargo package` work with `build.build-dir`
* Added templating support for `build.build-dir`
* Cleanup the commits a bit and tried to follow the request in [this comment](https://github.com/rust-lang/cargo/pull/15104#discussion_r1930859849) regarding tests

I updated the PR description with more details as well. notably:

* Added another question I would like input form the Cargo team on.
  *  Are bins generated from benches considered final artifacts? (see PR desc for more details)
* Updated the TODO list with a more detailed list of tasks before marking this PR as ready for review.
* Cleaned up the "What is a final artifact vs an intermediate build artifact" section
  * Also added the following file types to the list of intermediate build artifacts
    *  "pre-uplifted" binary executables. (ie. bins for `examples` that contain the hash in the name)
    * `CARGO_TARGET_TMPDIR` files (see rational for this [here](https://github.com/rust-lang/cargo/issues/14125#issuecomment-2258708917))

epage commented:
> Added templating support for build.build-dir

I haven't looked but would it work to split that out into a follow up PR?  Keeping PRs focused can help with ensuring a smooth review and that we are less likely to overlook something.

> Are bins generated from benches considered final artifacts? (see PR desc for more details)

My take is that if it isn't staged/uplifted, then its not a final artifact.

I assume this also applies to test binaries, build scripts, and proc-macros.

ranger-ross commented:
> I haven't looked but would it work to split that out into a follow up PR? Keeping PRs focused can help with ensuring a smooth review and that we are less likely to overlook something.

Sure. Its not a ton of code, but I have no problems breaking that out into anther PR to reduce the scope of this one :+1:

ranger-ross commented:
> > Are bins generated from benches considered final artifacts? (see PR desc for more details)
> 
> My take is that if it isn't staged/uplifted, then its not a final artifact.
> 
> I assume this also applies to test binaries, build scripts, and proc-macros.

Thanks, I think that sounds reasonable.
I will add that to the classification list in PR description with this as well.

epage commented on tests/testsuite/build_dir.rs:
```diff
@@ -0,0 +1,193 @@
+
+#[track_caller]
+fn assert_executeable_exists(path: PathBuf, name: &str, exists: bool) {
+    let executable = if cfg!(target_os = "windows") {
+        format!("{name}.exe")
+    } else {
+        name.to_string()
+    };
```
You can always insert https://doc.rust-lang.org/std/env/consts/constant.EXE_SUFFIX.html, making this not have to deal with conditional compilation

epage commented on tests/testsuite/build_dir.rs:
```diff
@@ -0,0 +1,193 @@
+        name.to_string()
+    };
+
+    assert_eq!(path.join(executable).exists(), exists);
```
Can you emit the path in the assert message?

epage commented on tests/testsuite/build_dir.rs:
```diff
@@ -96,51 +104,79 @@ fn cargo_package_should_output_to_target_dir() {
+}
+
+#[cargo_test]
+fn verify_build_dir_is_disabled_by_feature_flag() {
```
Can you add this test to the first commit?

epage commented on tests/testsuite/build_dir.rs:
```diff
@@ -96,51 +104,79 @@ fn cargo_package_should_output_to_target_dir() {
+    assert_build_dir(p.root().join("target"), "debug", true);
+    assert_executeable_exists(p.root().join("target/debug"), "foo", true);
     assert!(!p.root().join("build").exists());
 }
```
Looks like you are taking the approach "Ignore the feature that is used without a gate " (see https://doc.rust-lang.org/nightly/nightly-rustc/cargo/core/features/index.html)

Could you speak to that (and update the PR description with that)

ranger-ross replied:
Sure, I updated the PR description with a feature gating section

epage commented on src/cargo/util/context/mod.rs:
```diff
@@ -634,6 +634,35 @@ impl GlobalContext {
+            return self.target_dir();
+        }
+
+        if let Some(val) = &self.build_config()?.build_dir {
+            let path = val.resolve_path(self);
```
Compared to `pub fn target_dir()`,
- We don't have `self.build_dir` which is fine as we don't have a CLI flag
- We don't check `CARGO_BUILD_DIR` env variable as a shortcut for `CARGO_BUIILD_BUILD_DIR`.  **Should we have this shortcut?**

epage commented on src/cargo/util/context/mod.rs:
```diff
@@ -634,6 +634,35 @@ impl GlobalContext {
+            }
+
+            Ok(Some(Filesystem::new(path)))
+        } else {
+            // For now, fallback to the previous implementation.
+            // This will change in the future.
+            return self.target_dir();
+        }
```
Should we handle this fallback here or in `Workspace::build_dir`?

(I can see both ways)

epage replied:
Looks like it is in the follow up commit, duplicating the fallback

ranger-ross replied:
yes, I added the fallback in both places "just in case"

epage commented on src/cargo/core/compiler/fingerprint/mod.rs:
```diff
@@ -1857,7 +1857,7 @@ pub fn dep_info_loc(build_runner: &mut BuildRunner<'_, '_>, unit: &Unit) -> Path
 /// Returns an absolute path that target directory.
 /// All paths are rewritten to be relative to this.
 fn target_root(build_runner: &BuildRunner<'_, '_>) -> PathBuf {
-    build_runner.bcx.ws.target_dir().into_path_unlocked()
+    build_runner.bcx.ws.build_dir().into_path_unlocked()
```
Should `target_root` be renamed?

epage commented on src/cargo/core/compiler/future_incompat.rs:
```diff
@@ -166,7 +166,7 @@ impl OnDiskReports {
         let on_disk = serde_json::to_vec(&self).unwrap();
         if let Err(e) = ws
-            .target_dir()
+            .build_dir()
             .open_rw_exclusive_create(
                 FUTURE_INCOMPAT_FILE,
```
I'm not seeing this discussed in the PR as intermediate vs final.

ranger-ross replied:
Thanks for pointing this out. I reverted the change and added the following to the "final artifacts" category 
> * reports generated from `cargo report` like [future-incompat-report](https://doc.rust-lang.org/cargo/reference/future-incompat-report.html)

epage replied:
From those docs, it sounds like the user doesn't interact directly with the files but views them only through a command.   That sounds more like an intermediate artifact.

ranger-ross replied:
Hmm good point. Since it is JSON I assumed it was a user facing file. But looking at the docs, they do not reference that file so i guess its meant to be internal. 

I think in that cause it should belong in the build-dir.
I updated that list in the PR description and will revert the code for this.

---

Also, one thing that just came to my mind, is the HTML report of the `--timings` flag in Cargo.
I added this to the final artifacts list in the PR description.

epage commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -297,7 +297,7 @@ fn rustc(
 
     let root_output = build_runner.files().host_dest().to_path_buf();
-    let target_dir = build_runner.bcx.ws.target_dir().into_path_unlocked();
+    let target_dir = build_runner.bcx.ws.build_dir().into_path_unlocked();
```
Should we rename this variable?

epage commented on tests/testsuite/build_dir.rs:
```diff
@@ -12,19 +12,22 @@ fn binary_with_debug() {
     assert_build_dir(p.root().join("build"), "debug", true);
-    assert_executeable_exists(p.root().join("build/debug"), "foo", true);
-    assert!(!p.root().join("target").exists());
+    assert_build_dir(p.root().join("target"), "debug", false);
```
In reading these tests, the intent of the assertion isn't too clear (asserting that the target dir is a build dir? what does `false` mean?).  I think it'd be better to have two separate functions.  If they truly must share an implementation, then they can just be wrappers on top of that.

epage commented on src/cargo/ops/cargo_package/verify.rs:
```diff
@@ -45,9 +45,14 @@ pub fn run_verify(
-    let dst = tar
-        .parent()
-        .join(&format!("{}-{}", pkg.name(), pkg.version()));
+    let dst = if gctx.cli_unstable().build_dir {
+        ws.build_dir()
+            .as_path_unlocked()
+            .join(&format!("package/{}-{}", pkg.name(), pkg.version()))
+    } else {
+        tar.parent()
+            .join(&format!("{}-{}", pkg.name(), pkg.version()))
+    };
```
Why make this conditional?

ranger-ross replied:
Good point, looking a bit closer, it appears these should always be equal since the `tar` should [always be placed in `target/package`](https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_package/mod.rs#L104).

I'll simplify this to just 
```rs
let dst = ws.build_dir()
            .as_path_unlocked()
            .join(&format!("package/{}-{}", pkg.name(), pkg.version()))
```

epage commented on src/doc/src/reference/unstable.md:
````diff
@@ -237,6 +238,27 @@ This can also be specified in `.cargo/config.toml` files.
 
+## build-dir
+* Original Issue: [#14125](https://github.com/rust-lang/cargo/issues/14125)
+* Tracking Issue: None
+
+The directory where intermediate build artifacts will be stored.
+Intermediate artifacts are produced by Rustc/Cargo during the build process.
+
+* Type: string (path)
+* Default: Defaults to the value of `build.target-dir`
+* Environment: `CARGO_BUILD_BUILD_DIR` or `CARGO_BUILD_DIR`
+
+
+This can be specified in `.cargo/config.toml` files using the `build.build-dir`.
+
+Example:
+
+```toml
+[build]
+build-dir = "out"
+```
````
````suggestion
* Tracking Issue: [#14125](https://github.com/rust-lang/cargo/issues/14125)

The directory where intermediate build artifacts will be stored.
Intermediate artifacts are produced by Rustc/Cargo during the build process.

```toml
[build]
build-dir = "out"
```

### `build.build-dir`

* Type: string (path)
* Default: Defaults to the value of `build.target-dir`
* Environment: `CARGO_BUILD_BUILD_DIR` or `CARGO_BUILD_DIR`

The path to where internal files used as part of the build are placed.

````

ranger-ross commented on tests/testsuite/build_dir.rs:
Sorry, I think I misunderstood your request. (perhaps multiple times)

I added a doc comment at the [top of `build_dir.rs`](https://github.com/rust-lang/cargo/pull/15104/commits/b43e226b8039f1140d381e19790a9d651548b6d6#diff-d37a8e4e98440be7f035daa36682456743a69ede1307f7e189f219179725bf80R3-R10) with a summary of the test strategy for that file and the rational why.

Let me know if that is what you were asking for.

epage replied:
In the ideal scenario, we verify
- each file listed under artifact-dir vs build-dir is in the right place
- ensure new files get categorized correctly

I'm assuming we can't meet that ideal.  I would like to understand why we can't meet that ideal and understand how close we should get to it along with a call out of what gaps are left.

ranger-ross replied:
> I'm assuming we can't meet that ideal. I would like to understand why we can't meet that ideal and understand how close we should get to it along with a call out of what gaps are left.

I spent some time adding some additional tests to get a bit closer to this ideal.
I tried to have a test for every bullet point in the `build-dir`/`artifact-dir` list in the PR description. 

Gaps
* Not all files are explicitly checked for various reason.
  * `cargo doc` output contains a lot of files and I suspect the files are subject to change as rustdoc changes which will make this a maintenance burden.
  * `incremental build` file names are just the hash so it is a bit difficult to check explicitly. I settled for simply checking for `incremental` directory in the build-dir.
* Hashes in the file/dir names is a bit tedious to test with the current approach (primarily using `glob` crate).
  * I am currently using a `*` wildcard, but this leaves 
  * It should be possible to improve this in the future (I am thinking by using `regex` and `walkdir` crates together)
* I am not a cargo expert so there may be other files I could have missed that I am just not aware of.

epage replied:
Agreed that we shouldn't need to check every individual file generated but each class of artifact file (where does `cargo doc` get written to generally, where does incremental builds get written to, etc)

epage replied:
Opened more specific comments.  Thanks for exploring this!

weihanglo commented on src/cargo/util/context/mod.rs:
```diff
@@ -634,6 +634,44 @@ impl GlobalContext {
+        if !self.cli_unstable().build_dir {
+            return self.target_dir();
+        }
+        if let Some(dir) = self.get_env_os("CARGO_BUILD_DIR") {
```
I understand `CARGO_BUILD_DIR` is proposed here because we already have `CARGO_TARGET_DIR`. However I feel like `CARGO_TARGET_DIR` was a mistake.

The `build.build-dir` config already comes with a `CARGO_BUILD_BUILD_DIR` environment variable for free. We may want to stick with it and stop stacking up extra environment variables.

ranger-ross replied:
> However I feel like `CARGO_TARGET_DIR` was a mistake.

Can I ask why you feel it was a mistake? 

As a user of cargo I think it might be a bit confusing to see `BUILD` show up twice in the same variable name.
It make sense when strictly following the [Cargo configuration](https://doc.rust-lang.org/cargo/reference/config.html#environment-variables), but newcomers are often not aware of this.

I think there is value in providing a shortcut that is easily understandable.

epage replied:
> And the --target-dir flag can just be a alias equivalent to --config 'build.target-dir=foo', and could eliminate the special handling.

The plan was to not offer a `--build-dir` env variable and build-dir would eventually be decoupled from `--target-dir`

ranger-ross replied:
> Could you elaborate a bit more on this? In the book there is a list of variables for each configuration field, which I assume no discovery issue here. (Could expand the doc of CARGO_BUILD_TARGET_DIR, if the other one is removed, yet we won't remove it for sure.)

I think that anyone remotely familiar with Rust can fairly quickly understand `CARGO_TARGET_DIR` does without going to the documentation page. Seeing `CARGO_BUILD_BUILD_DIR` is much less obvious and will probably require some google searching to figure out what it does. 

But perhaps to your later point, if people just set it once in their shared cargo config, perhaps it does not matter that much.

> While I lean towards `CARGO_BUILD_DIR`, starting with fewer features makes it easier to identify when a feature is needed compared to when you offer everything, its hard to tell what is used and could be removed.

I think this is fair. I can remove it from this PR and we can revisit if it's really needed later

weihanglo commented on src/cargo/core/compiler/layout.rs:
```diff
@@ -150,15 +157,22 @@ impl Layout {
         //
         paths::create_dir_all_excluded_from_backups_atomic(root.as_path_unlocked())?;
+        if root != build_root {
+            paths::create_dir_all_excluded_from_backups_atomic(build_root.as_path_unlocked())?;
```
Not really on topic, but this reminds me <https://github.com/rust-lang/cargo/issues/11548> and <https://github.com/rust-lang/cargo/issues/15061>.

For the new build-dir, we don't have backward compatibility issue, and files inside are really intermediate caches. We might want to reconsider the self-ignoring directory approach.

epage commented on tests/testsuite/build_dir.rs:
```diff
@@ -0,0 +1,589 @@
+        &[
+            &format!("build/debug/deps/foo*{EXE_SUFFIX}"),
+            "build/debug/deps/foo*.d",
+            "build/debug/.fingerprint/foo-*/bin-foo",
+            "build/debug/.fingerprint/foo-*/bin-foo.json",
+            "build/debug/.fingerprint/foo-*/dep-bin-foo",
+            "build/debug/.fingerprint/foo-*/invoked.timestamp",
```
We probably only need to check where `.fingerprint/` is written to, not every class of file inside of it

epage commented on tests/testsuite/build_dir.rs:
```diff
@@ -0,0 +1,589 @@
+}
+
+#[cargo_test]
+fn binary_with_release() {
```
I feel like this should be next to `binary_with_debug` as they seem to be closely related

epage commented on tests/testsuite/build_dir.rs:
```diff
@@ -0,0 +1,589 @@
+    assert_exists_patterns_with_base_dir(
+        &p.root(),
+        &[
+            &format!("build/debug/build/foo-*/build-script-build{EXE_SUFFIX}"),
+            "build/debug/build/foo-*/build_script_build*.d",
+            "build/debug/build/foo-*/invoked.timestamp",
+            "build/debug/build/foo-*/out",
+            "build/debug/build/foo-*/out/foo.txt", // Verify OUT_DIR
+            "build/debug/build/foo-*/output",
+            "build/debug/build/foo-*/root-output",
+            "build/debug/build/foo-*/stderr",
```
Similarly, do we need to check every build script file or just `build/` exists in only one location?

Hmm,. the `build/*/build` repetition could make this confusing.  Should we have the tests call the first one `build_dir`?

ranger-ross replied:
> Hmm,. the build/*/build repetition could make this confusing. Should we have the tests call the first one build_dir?

Yeah that makes sense. Also I think I might have made the review commits a bit confusing by naming `target-dir = "build"` in the first test commit and then renamed it in the impl commit. :sweat_smile: 

I changed it to `target-dir = "target-dir"` and only added `build-dir = "build-dir"` in the impl commit.
I think this makes the git diffs much easier to understand.

epage commented on tests/testsuite/build_dir.rs:
```diff
@@ -47,17 +47,19 @@ fn binary_with_debug() {
 
-    p.cargo("build")
-        .masquerade_as_nightly_cargo(&[])
+    p.cargo("build -Z unstable-options -Z build-dir")
```
why is `-Zunstable-options` used?

ranger-ross replied:
oops, this is from a copy/paste from the `--artfiact-dir` tests. Since build-dir does not a CLI option its not needed. 
Removed it :+1:

epage commented on tests/testsuite/build_dir.rs:
```diff
@@ -0,0 +1,589 @@
+}
+
+#[cargo_test]
+fn build_script_should_output_to_build_dir() {
```
So this is named in the test commit for what the behavior should be while `cargo_tmpdir_should_output_to_target_dir` is named for the current behavior but not renamed while `examples_should_output_to_target_dir` is named for the current behavior and then renamed when the behavior changed (as examples)

We should probably be consistent and ensure the names are correct

epage replied:
I think "making the names either consistently renamed or consistently always the final behavior" is the only thing left?

ranger-ross replied:
Sorry, I totally forgot to up the test names in the second commit :sweat_smile: 

Updated them in https://github.com/rust-lang/cargo/compare/dca7eb511c..5874b6f3474a

ranger-ross commented:
Rebased to pick up the CI fixes in #15222 to hopefully make CI green

epage commented:
Thanks for putting in all this work on this!
